File: sdk/go/backend_groups.go
package tempemail

import (
	"sync"
	"time"
)

var backendGroups = map[string][]Channel{
	"mailinator": {
		ChannelMailinator,
		"sogetthis-com", "bobmail-info", "suremail-info", "binkmail-com",
		"veryrealemail-com", "chammy-info", "thisisnotmyrealemail-com",
		"notmailinator-com", "spamhereplease-com", "sendspamhere-com",
		"sendfree-org",
	},
	"getnada": {
		ChannelGetnada,
		"getnada-com", "getnada-email", "getnada-net",
	},
	"guerrillamail": {
		ChannelGuerrillaMail,
		"sharklasers",
	},
	"moakt": {
		ChannelMoakt,
		"drmail-in", "teml-net", "tmpeml-com", "tmpbox-net",
		"moakt-cc", "disbox-net", "tmpmail-org", "tmpmail-net",
		"tmails-net", "disbox-org", "moakt-co", "moakt-ws",
		"tmail-ws", "bareed-ws",
	},
	"catchmail": {
		ChannelCatchmail,
		"catchmail-mailistry", "catchmail-zeppost",
	},
	"mailforspam": {
		ChannelMailforspam,
		"mailforspam-tempmail-io", "mailforspam-disposable",
	},
	"tempmail-plus": {
		ChannelTempmailPlus,
		"fexpost-com", "fexbox-org", "mailbox-in-ua",
		"rover-info", "chitthi-in", "fextemp-com",
		"any-pink", "merepost-com",
	},
	"mail-cx": {
		ChannelMailCx,
		"ddker-com",
	},
	"fake-legal": {
		ChannelFakeLegal,
		"imgui-de", "pulsewebmenu-de",
	},
	"neighbours": {
		"neighbours-sh", "neighbours",
	},
}

var channelToBackend map[Channel]string

func init() {
	channelToBackend = make(map[Channel]string)
	for backend, channels := range backendGroups {
		for _, ch := range channels {
			channelToBackend[ch] = backend
		}
	}
}

type circuitState struct {
	lastFailure time.Time
	failCount   int
}

var (
	circuitMu  sync.Mutex
	circuitMap = make(map[string]*circuitState)
)

const defaultCooldown = 60 * time.Second
const maxCooldown = 5 * time.Minute

func isBackendOpen(backend string) bool {
	circuitMu.Lock()
	defer circuitMu.Unlock()
	state, ok := circuitMap[backend]
	if !ok {
		return true
	}
	cooldown := defaultCooldown * time.Duration(1<<(state.failCount-1))
	if cooldown > maxCooldown {
		cooldown = maxCooldown
	}
	if time.Since(state.lastFailure) >= cooldown {
		delete(circuitMap, backend)
		return true
	}
	return false
}

func recordBackendFailure(backend string) {
	circuitMu.Lock()
	defer circuitMu.Unlock()
	state, ok := circuitMap[backend]
	if ok {
		state.lastFailure = time.Now()
		state.failCount++
	} else {
		circuitMap[backend] = &circuitState{lastFailure: time.Now(), failCount: 1}
	}
}

func recordBackendSuccess(backend string) {
	circuitMu.Lock()
	defer circuitMu.Unlock()
	delete(circuitMap, backend)
}

File: sdk/rust/src/backend_groups.rs
use std::collections::HashMap;
use std::sync::{LazyLock, Mutex};
use std::time::Instant;

use crate::types::Channel;

static BACKEND_GROUPS: LazyLock<HashMap<&'static str, Vec<Channel>>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    m.insert("mailinator", vec![
        Channel::Mailinator,
        Channel::SogetthisCom,
        Channel::BobmailInfo,
        Channel::SuremailInfo,
        Channel::BinkmailCom,
        Channel::VeryrealemailCom,
        Channel::ChammyInfo,
        Channel::ThisisnotmyrealemailCom,
        Channel::NotmailinatorCom,
        Channel::SpamherepleaseCom,
        Channel::SendspamhereCom,
        Channel::SendfreeOrg,
    ]);
    m.insert("getnada", vec![
        Channel::Getnada,
        Channel::GetnadaCom,
        Channel::GetnadaEmail,
        Channel::GetnadaNet,
    ]);
    m.insert("guerrillamail", vec![
        Channel::GuerrillaMail,
        Channel::Sharklasers,
    ]);
    m.insert("moakt", vec![
        Channel::Moakt,
        Channel::DrmailIn,
        Channel::TemlNet,
        Channel::TmpemlCom,
        Channel::TmpboxNet,
        Channel::MoaktCc,
        Channel::DisboxNet,
        Channel::TmpmailOrg,
        Channel::TmpmailNet,
        Channel::TmailsNet,
        Channel::DisboxOrg,
        Channel::MoaktCo,
        Channel::MoaktWs,
        Channel::TmailWs,
        Channel::BareedWs,
    ]);
    m.insert("catchmail", vec![
        Channel::Catchmail,
        Channel::CatchmailMailistry,
        Channel::CatchmailZeppost,
    ]);
    m.insert("mailforspam", vec![
        Channel::Mailforspam,
        Channel::MailforspamTempmailIo,
        Channel::MailforspamDisposable,
    ]);
    m.insert("tempmail-plus", vec![
        Channel::TempmailPlus,
        Channel::FexpostCom,
        Channel::FexboxOrg,
        Channel::MailboxInUa,
        Channel::RoverInfo,
        Channel::ChitthiIn,
        Channel::FextempCom,
        Channel::AnyPink,
        Channel::MerepostCom,
    ]);
    m.insert("mail-cx", vec![
        Channel::MailCx,
        Channel::DdkerCom,
    ]);
    m.insert("fake-legal", vec![
        Channel::FakeLegal,
        Channel::ImguiDe,
        Channel::PulsewebmenuDe,
    ]);
    m.insert("neighbours", vec![
        Channel::NeighboursSh,
        Channel::Neighbours,
    ]);
    m
});

static CHANNEL_TO_BACKEND: LazyLock<HashMap<Channel, &'static str>> = LazyLock::new(|| {
    let mut m = HashMap::new();
    for (backend, channels) in BACKEND_GROUPS.iter() {
        for ch in channels {
            m.insert(ch.clone(), *backend);
        }
    }
    m
});

pub fn get_backend(channel: &Channel) -> Option<&'static str> {
    CHANNEL_TO_BACKEND.get(channel).copied()
}

const DEFAULT_COOLDOWN_SECS: f64 = 60.0;
const MAX_COOLDOWN_SECS: f64 = 300.0;

struct CircuitState {
    last_failure: Instant,
    fail_count: u32,
}

static CIRCUIT_MAP: LazyLock<Mutex<HashMap<&'static str, CircuitState>>> =
    LazyLock::new(|| Mutex::new(HashMap::new()));

pub fn is_backend_open(backend: &str) -> bool {
    let mut map = CIRCUIT_MAP.lock().unwrap();
    if let Some(state) = map.get(backend) {
        let cooldown_secs =
            (DEFAULT_COOLDOWN_SECS * 2f64.powi((state.fail_count as i32) - 1)).min(MAX_COOLDOWN_SECS);
        let elapsed = state.last_failure.elapsed().as_secs_f64();
        if elapsed >= cooldown_secs {
            map.remove(backend);
            true
        } else {
            false
        }
    } else {
        true
    }
}

pub fn record_backend_failure(backend: &'static str) {
    let mut map = CIRCUIT_MAP.lock().unwrap();
    if let Some(state) = map.get_mut(backend) {
        state.last_failure = Instant::now();
        state.fail_count += 1;
    } else {
        map.insert(backend, CircuitState {
            last_failure: Instant::now(),
            fail_count: 1,
        });
    }
}

pub fn record_backend_success(backend: &str) {
    let mut map = CIRCUIT_MAP.lock().unwrap();
    map.remove(backend);
}
